lib/trino-filesystem-cache-alluxio/src/main/java/io/trino/filesystem/alluxio/AlluxioFileSystemCacheModule.java

```diff
@@ -0,0 +1,115 @@
+import static java.lang.String.join;
+import static org.weakref.jmx.guice.ExportBinder.newExporter;
+
+public class AlluxioFileSystemCacheModule
+        extends AbstractConfigurationAwareModule
+{
+    @Override
+    protected void setup(Binder binder)
+    {
+        configBinder(binder).bindConfig(AlluxioFileSystemCacheConfig.class);
+        binder.bind(CacheStats.class).to(AlluxioCacheStats.class).in(SINGLETON);
+        newExporter(binder).export(CacheStats.class).as(generator -> generator.generatedNameOf(AlluxioCacheStats.class));
+
+        binder.bind(TrinoFileSystemCache.class).to(AlluxioFileSystemCache.class).in(SINGLETON);
+
+        Properties metricProps = new Properties();
+        metricProps.put("sink.jmx.class", "alluxio.metrics.sink.JmxSink");
+        metricProps.put("sink.jmx.domain", "org.alluxio");
+        MetricsSystem.startSinksFromConfig(new MetricsConfig(metricProps));
+    }
+
+    @Provides
+    @Singleton
+    public static AlluxioConfiguration getAlluxioConfiguration(AlluxioFileSystemCacheConfig config)
+    {
+        checkArgument(config.getMaxCacheSizes().isEmpty() ^ config.getMaxCacheDiskUsagePercentages().isEmpty(),
+                "Either %s or %s must be specified", CACHE_MAX_SIZES, CACHE_MAX_PERCENTAGES);
+        int size = config.getMaxCacheSizes().isEmpty() ? config.getMaxCacheDiskUsagePercentages().size() : config.getMaxCacheSizes().size();
+        checkArgument(config.getCacheDirectories().size() == size,
+                "%s and %s must have the same size", CACHE_DIRECTORIES, config.getMaxCacheSizes().isEmpty() ? CACHE_MAX_PERCENTAGES : CACHE_MAX_SIZES);
+        List<DataSize> maxCacheSizes = config.getMaxCacheSizes().isEmpty() ?
+                calculateMaxCacheSizes(config.getMaxCacheDiskUsagePercentages(), config.getCacheDirectories().stream()
+                        .map(directory -> new File(directory).getTotalSpace()).collect(toImmutableList()))
+                : config.getMaxCacheSizes();
+
+        AlluxioProperties alluxioProperties = new AlluxioProperties();
+        alluxioProperties.set(USER_CLIENT_CACHE_ENABLED, true);
+        alluxioProperties.set(USER_CLIENT_CACHE_DIRS, join(",", config.getCacheDirectories()));
+        alluxioProperties.set(USER_CLIENT_CACHE_SIZE, join(",", maxCacheSizes.stream().map(DataSize::toBytesValueString).toList()));
+        alluxioProperties.set(USER_CLIENT_CACHE_PAGE_SIZE, config.getCachePageSize().toBytesValueString());
+        alluxioProperties.set(USER_CLIENT_CACHE_TTL_THRESHOLD_SECONDS, config.getCacheTTL().roundTo(TimeUnit.SECONDS));
+        alluxioProperties.set(USER_CLIENT_CACHE_TTL_ENABLED, true);
+        return new InstancedConfiguration(alluxioProperties);
+    }
+
+    @Provides
+    @Singleton
+    public static CacheManager getCacheManager(AlluxioConfiguration alluxioConfiguration)
+    {
+        try {
+            return CacheManager.Factory.create(alluxioConfiguration);
```

Comment:
I tried configuring this feature on a machine where the cache directory did not exist. The configuration error wasn't caught and Trino server startup succeeded, all queries on the caching configured catalog fail.
```
2024-01-02T06:43:54.959Z	INFO	main	alluxio.client.file.cache.PageStore	Opening PageStore with option=alluxio.client.file.cache.store.PageStoreOptions@15b078f
2024-01-02T06:43:54.993Z	ERROR	pool-48-thread-1	alluxio.client.file.cache.LocalCacheManager	IOException occurs in page scan
java.nio.file.NoSuchFileException: /opt/data/LOCAL
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55)
	at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:171)
	at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99)
	at java.base/java.nio.file.Files.readAttributes(Files.java:1853)
	at java.base/java.nio.file.FileTreeWalker.getAttributes(FileTreeWalker.java:220)
	at java.base/java.nio.file.FileTreeWalker.visit(FileTreeWalker.java:277)
	at java.base/java.nio.file.FileTreeWalker.walk(FileTreeWalker.java:323)
	at java.base/java.nio.file.FileTreeIterator.<init>(FileTreeIterator.java:71)
	at java.base/java.nio.file.Files.walk(Files.java:3898)
	at java.base/java.nio.file.Files.walk(Files.java:3953)
	at alluxio.client.file.cache.store.LocalPageStoreDir.scanPages(LocalPageStoreDir.java:98)
	at alluxio.client.file.cache.LocalCacheManager.lambda$invalidate$8(LocalCacheManager.java:870)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:422)
	at alluxio.client.file.cache.LocalCacheManager.invalidate(LocalCacheManager.java:868)
	at alluxio.client.file.cache.LocalCacheManager.lambda$new$2(LocalCacheManager.java:167)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)


2024-01-02T06:43:54.996Z	INFO	pool-47-thread-1	alluxio.client.file.cache.LocalCacheManager	Restoring PageStoreDir (/opt/data/LOCAL)
2024-01-02T06:43:54.996Z	ERROR	pool-47-thread-1	alluxio.client.file.cache.LocalCacheManager	Failed to restore PageStore: Directory /opt/data/LOCAL does not exist
2024-01-02T06:43:55.007Z	ERROR	pool-47-thread-1	alluxio.client.file.cache.LocalCacheManager	Cache is in NOT_IN_USE.
2024-01-02T06:43:55.007Z	ERROR	pool-47-thread-1	alluxio.client.file.cache.LocalCacheManager	Failed to restore LocalCacheManager
java.nio.file.AccessDeniedException: /opt/data
	at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:90)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
	at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
	at java.base/sun.nio.fs.UnixFileSystemProvider.createDirectory(UnixFileSystemProvider.java:438)
	at java.base/java.nio.file.Files.createDirectory(Files.java:699)
	at java.base/java.nio.file.Files.createAndCheckIsDirectory(Files.java:807)
	at java.base/java.nio.file.Files.createDirectories(Files.java:793)
	at alluxio.client.file.cache.store.PageStoreDir.clear(PageStoreDir.java:97)
	at alluxio.client.file.cache.store.LocalPageStoreDir.reset(LocalPageStoreDir.java:86)
	at alluxio.client.file.cache.LocalCacheManager.restoreOrInit(LocalCacheManager.java:759)
	at alluxio.client.file.cache.LocalCacheManager.lambda$create$0(LocalCacheManager.java:130)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:317)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)


2024-01-02T06:43:56.917Z	INFO	main	org.ishugaliy.allgood.consistent.hash.HashRing	Ring [hash_ring_1642] created: hasher [METRO_HASH], partitionRate [1000]
2024-01-02T06:43:57.118Z	INFO	main	io.trino.connector.StaticCatalogManager	-- Added catalog delta_lake using connector delta_lake --
```
We should ensure that either we create the configured cache directory or fail server startup if we can't. Currently it looks like configuration failure is getting swallowed somewhere.

Reply:
Once I got past above configuration problem, I've run into a situation where one of my nodes is caching correctly while the other is causing failures
```
io.trino.spi.TrinoException: Failed to read Parquet file: s3://starburst-benchmarks-data/delta_lake-tpcds-sf1000-PARQUET-part/item-be94bb8a1672454b937d6f760b493911/20230630_145116_00979_wcd7b-128edf22-6262-41e9-bdc6-f353e849a4eb
	at io.trino.plugin.hive.parquet.ParquetPageSource.handleException(ParquetPageSource.java:206)
	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.lambda$createPageSource$1(ParquetPageSourceFactory.java:277)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:75)
	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:361)
	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:340)
	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:235)
	at io.trino.spi.Page.getLoadedPage(Page.java:237)
	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:269)
	at io.trino.operator.Driver.processInternal(Driver.java:398)
	at io.trino.operator.Driver.lambda$process$8(Driver.java:301)
	at io.trino.operator.Driver.tryWithLock(Driver.java:704)
	at io.trino.operator.Driver.process(Driver.java:293)
	at io.trino.operator.Driver.processForDuration(Driver.java:264)
	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
	at io.trino.$gen.Trino_94a8d06____20240102_093407_2.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: java.lang.RuntimeException: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:109)
	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:42)
	at com.google.common.collect.Iterators$PeekingImpl.peek(Iterators.java:1218)
	at io.trino.parquet.reader.PageReader.readDictionaryPage(PageReader.java:160)
	at io.trino.parquet.reader.AbstractColumnReader.setPageReader(AbstractColumnReader.java:79)
	at io.trino.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:457)
	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:551)
	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:534)
	at io.trino.parquet.reader.ParquetReader.lambda$nextPage$3(ParquetReader.java:251)
	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:72)
	... 17 more
Caused by: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
	at org.apache.parquet.format.Util.read(Util.java:366)
	at org.apache.parquet.format.Util.readPageHeader(Util.java:133)
	at org.apache.parquet.format.Util.readPageHeader(Util.java:128)
	at io.trino.parquet.reader.ParquetColumnChunkIterator.readPageHeader(ParquetColumnChunkIterator.java:116)
	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:83)
	... 26 more
Caused by: shaded.parquet.org.apache.thrift.protocol.TProtocolException: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1114)
	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1025)
	at org.apache.parquet.format.PageHeader.read(PageHeader.java:902)
	at org.apache.parquet.format.Util.read(Util.java:363)
	... 30 more
```
Is it possible that we're swallowing failures to write into cache somewhere ?

Reply:
> Once I got past above configuration problem, I've run into a situation where one of my nodes is caching correctly while the other is causing failures
> 
> ```
> io.trino.spi.TrinoException: Failed to read Parquet file: s3://starburst-benchmarks-data/delta_lake-tpcds-sf1000-PARQUET-part/item-be94bb8a1672454b937d6f760b493911/20230630_145116_00979_wcd7b-128edf22-6262-41e9-bdc6-f353e849a4eb
> 	at io.trino.plugin.hive.parquet.ParquetPageSource.handleException(ParquetPageSource.java:206)
> 	at io.trino.plugin.hive.parquet.ParquetPageSourceFactory.lambda$createPageSource$1(ParquetPageSourceFactory.java:277)
> 	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:75)
> 	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:361)
> 	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:340)
> 	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:235)
> 	at io.trino.spi.Page.getLoadedPage(Page.java:237)
> 	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:269)
> 	at io.trino.operator.Driver.processInternal(Driver.java:398)
> 	at io.trino.operator.Driver.lambda$process$8(Driver.java:301)
> 	at io.trino.operator.Driver.tryWithLock(Driver.java:704)
> 	at io.trino.operator.Driver.process(Driver.java:293)
> 	at io.trino.operator.Driver.processForDuration(Driver.java:264)
> 	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
> 	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
> 	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
> 	at io.trino.$gen.Trino_94a8d06____20240102_093407_2.run(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
> 	at java.base/java.lang.Thread.run(Thread.java:1583)
> Caused by: java.lang.RuntimeException: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:109)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:42)
> 	at com.google.common.collect.Iterators$PeekingImpl.peek(Iterators.java:1218)
> 	at io.trino.parquet.reader.PageReader.readDictionaryPage(PageReader.java:160)
> 	at io.trino.parquet.reader.AbstractColumnReader.setPageReader(AbstractColumnReader.java:79)
> 	at io.trino.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:457)
> 	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:551)
> 	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:534)
> 	at io.trino.parquet.reader.ParquetReader.lambda$nextPage$3(ParquetReader.java:251)
> 	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:72)
> 	... 17 more
> Caused by: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
> 	at org.apache.parquet.format.Util.read(Util.java:366)
> 	at org.apache.parquet.format.Util.readPageHeader(Util.java:133)
> 	at org.apache.parquet.format.Util.readPageHeader(Util.java:128)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.readPageHeader(ParquetColumnChunkIterator.java:116)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:83)
> 	... 26 more
> Caused by: shaded.parquet.org.apache.thrift.protocol.TProtocolException: Required field 'uncompressed_page_size' was not found in serialized data! Struct: org.apache.parquet.format.PageHeader$PageHeaderStandardScheme@30d6a9c4
> 	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1114)
> 	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1025)
> 	at org.apache.parquet.format.PageHeader.read(PageHeader.java:902)
> 	at org.apache.parquet.format.Util.read(Util.java:363)
> 	... 30 more
> ```
> 
> Is it possible that we're swallowing failures to write into cache somewhere ?

thanks for all the hardwork authors and we are really excited about this change. So we did test it out by extending it to iceberg - I see files being cached at Workers however workers seem to fail in the process with different stack traces. Just putting one of them here for reference just in case it gives some clue to potential bug:

>  io.trino.spi.TrinoException: Failed to read Parquet file: s3a://acs-spark-core/tpcds/1000g-parquet-zstd-512mb-block-partitioned/customer/part-00000-e611ca3b-26c0-400a-bca1-519d62deedba-c000.zstd.parquet
> 	at io.trino.plugin.iceberg.IcebergPageSourceProvider.handleException(IcebergPageSourceProvider.java:1573)
> 	at io.trino.plugin.iceberg.IcebergPageSourceProvider.lambda$createParquetPageSource$26(IcebergPageSourceProvider.java:1143)
> 	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:75)
> 	at io.trino.spi.block.LazyBlock$LazyData.load(LazyBlock.java:405)
> 	at io.trino.spi.block.LazyBlock$LazyData.getFullyLoadedBlock(LazyBlock.java:384)
> 	at io.trino.spi.block.LazyBlock.getLoadedBlock(LazyBlock.java:291)
> 	at io.trino.spi.Page.getLoadedPage(Page.java:237)
> 	at io.trino.operator.TableScanOperator.getOutput(TableScanOperator.java:305)
> 	at io.trino.operator.Driver.processInternal(Driver.java:395)
> 	at io.trino.operator.Driver.lambda$process$8(Driver.java:298)
> 	at io.trino.operator.Driver.tryWithLock(Driver.java:694)
> 	at io.trino.operator.Driver.process(Driver.java:290)
> 	at io.trino.operator.Driver.processForDuration(Driver.java:261)
> 	at io.trino.execution.SqlTaskExecution$DriverSplitRunner.processFor(SqlTaskExecution.java:887)
> 	at io.trino.execution.executor.timesharing.PrioritizedSplitRunner.process(PrioritizedSplitRunner.java:187)
> 	at io.trino.execution.executor.timesharing.TimeSharingTaskExecutor$TaskRunner.run(TimeSharingTaskExecutor.java:565)
> 	at io.trino.$gen.Trino_430_61_gebf29ba_dirty____20240118_154324_2.run(Unknown Source)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
> 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
> 	at java.base/java.lang.Thread.run(Thread.java:840)
> Caused by: java.lang.RuntimeException: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: don't know what type: 13
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:143)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:47)
> 	at com.google.common.collect.Iterators$PeekingImpl.peek(Iterators.java:1218)
> 	at io.trino.parquet.reader.PageReader.readDictionaryPage(PageReader.java:186)
> 	at io.trino.parquet.reader.AbstractColumnReader.setPageReader(AbstractColumnReader.java:79)
> 	at io.trino.parquet.reader.ParquetReader.readPrimitive(ParquetReader.java:448)
> 	at io.trino.parquet.reader.ParquetReader.readColumnChunk(ParquetReader.java:546)
> 	at io.trino.parquet.reader.ParquetReader.readBlock(ParquetReader.java:529)
> 	at io.trino.parquet.reader.ParquetReader.lambda$nextPage$3(ParquetReader.java:279)
> 	at io.trino.parquet.reader.ParquetBlockFactory$ParquetBlockLoader.load(ParquetBlockFactory.java:72)
> 	... 17 more
> Caused by: java.io.IOException: can not read class org.apache.parquet.format.PageHeader: don't know what type: 13
> 	at org.apache.parquet.format.Util.read(Util.java:366)
> 	at org.apache.parquet.format.Util.readPageHeader(Util.java:133)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.readPageHeader(ParquetColumnChunkIterator.java:150)
> 	at io.trino.parquet.reader.ParquetColumnChunkIterator.next(ParquetColumnChunkIterator.java:116)
> 	... 26 more
> Caused by: shaded.parquet.org.apache.thrift.protocol.TProtocolException: don't know what type: 13
> 	at shaded.parquet.org.apache.thrift.protocol.TCompactProtocol.getTType(TCompactProtocol.java:899)
> 	at shaded.parquet.org.apache.thrift.protocol.TCompactProtocol.readListBegin(TCompactProtocol.java:598)
> 	at shaded.parquet.org.apache.thrift.protocol.TCompactProtocol.readSetBegin(TCompactProtocol.java:610)
> 	at org.apache.parquet.format.InterningProtocol.readSetBegin(InterningProtocol.java:179)
> 	at shaded.parquet.org.apache.thrift.protocol.TProtocolUtil.skip(TProtocolUtil.java:128)
> 	at shaded.parquet.org.apache.thrift.protocol.TProtocolUtil.skip(TProtocolUtil.java:112)
> 	at shaded.parquet.org.apache.thrift.protocol.TProtocolUtil.skip(TProtocolUtil.java:60)
> 	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1106)
> 	at org.apache.parquet.format.PageHeader$PageHeaderStandardScheme.read(PageHeader.java:1025)
> 	at org.apache.parquet.format.PageHeader.read(PageHeader.java:902)
> 	at org.apache.parquet.format.Util.read(Util.java:363)
> 	... 29 more

Reply:
Hey @amoghmargoor @sdaberdaku, thank you for giving it a go! I'm building this locally and will look into this, I assume something went wrong during the PR feedback/fixes cycle 👍

Reply:
There was a simple bug in the code for updating the cache. Whenever we fetched multiple pages from external storage we would set the first page as the cache entry for all the other pages. I've pushed a fix for it now. It could be reproduced in a test with:
```
String table = "test_table" + UUID.randomUUID().toString().replace("-", "_");
getQueryRunner().execute("CREATE TABLE "+ table + " AS SELECT * FROM tpch.sf1.customer limit 150000");
assertEquals(150000, getQueryRunner().execute("SELECT * FROM " + table).getRowCount());
```
It's a bit disappointing that none of the tests caught the bug.

Reply:
I've added an `@FileExists` annotation to the cache directory to hopefully catch issues with the cache directories when the configuration is validated.

Reply:
Do we have a test covering this now ?
I think we should also add a product test along the lines of TestHiveCaching and EnvMultinodeHiveCaching to make sure we have an end-to-end test and a working example of how to configure this.